### DISCUSSION TITLE
How are you meant to implement `discover_positions`?

### DISCUSSION THREAD

**Discussion by ColinKennedy** [nvim-neotest/neotest#385] [Q&A] [answered]
I have a C/CPP file the looks like this

```cpp
CTEST(suite_name, test_here1)
{
    ASSERT_TRUE(true);
}

CTEST(suite_name, test_here2)
{
    ASSERT_TRUE(false);
}
```

The namespace is `suite_name`, the test name is `testout`. I cannot figure out how to get a query going with `require("neotest.lib").lib.treesitter.parse_positions` to get the tests nested properly.

A proper tree should look like this

```
- test_file_name.cpp
    - suite_name
        - test_here1
        - test_here2
```

I tried writing a query like this

```lua
local lib = require("neotest.lib")

local M = {}

function M.discover_positions(path)
  local query = [[
    (function_definition
      declarator: (function_declarator
        declarator: (identifier)
        parameters: (parameter_list
          .
          (parameter_declaration
            type: (type_identifier) @namespace.name) @namespace.definition
          (parameter_declaration
            type: (type_identifier)))))

    (function_definition
      declarator: (function_declarator
        declarator: (identifier)
        parameters: (parameter_list
          .
          (parameter_declaration
            type: (type_identifier))
          (parameter_declaration
            type: (type_identifier) @test.name)))) @test.definition

  ]]

  return lib.treesitter.parse_positions(path, query, {
    require_namespaces = false,
  })
end


return M
```

But it creates a flattened tree like this:

```
- test_file_name.cpp
    - test_here1
    - test_here2
    - suite_name
```

I feel like I'm missing something simple. At the moment I'm close to saying "oh forget it" and writing the `neotest.Tree` myself since none of my attempts to read other neotest adapters so far have gotten me closer to figuring out how to get the right tree structure. If someone can point me in the right direction, I'd appreciate it. Thank you!

**Comment by rcarriga** [comment, marked as answer]
Unfortunately the treesitter library is just meant for helper methods, it's not guaranteed to work for everyone. The problem here is that it assumes a file structure where child positions are within the scope of the parent which is not going to be the case with how your tests are defined where the suite name is just an argument.

The gtest adapter had similar issues and it looks like they went with the manual tree building as well, maybe take a look and see if you can re-use some of their logic? https://github.com/alfaix/neotest-gtest/blob/main/lua/neotest-gtest/parse.lua

**Reply by ColinKennedy** [reply]
Thank you for your quick reply. I had a look and basically combined neotest-gtest and neotest-bash together to get what I needed. It'll still be a while before this is ready to share with others but it's going in a good direction again. Thank you!
